Tests/test_image_reduce.py

```diff
@@ -0,0 +1,244 @@
+from PIL import Image, ImageMath, ImageMode
+
+from .helper import PillowTestCase, convert_to_comparable
+
+
+class TestImageReduce(PillowTestCase):
+    # There are several internal implementations
+    remarkable_factors = [
+        # special implementations
+        1,
+        2,
+        3,
+        4,
+        5,
+        6,
+        # 1xN implementation
+        (1, 2),
+        (1, 3),
+        (1, 4),
+        (1, 7),
+        # Nx1 implementation
+        (2, 1),
+        (3, 1),
+        (4, 1),
+        (7, 1),
+        # general implementation with different paths
+        (4, 6),
+        (5, 6),
+        (4, 7),
+        (5, 7),
+        (19, 17),
+    ]
+
+    @classmethod
+    def setUpClass(cls):
+        cls.gradients_image = Image.open("Tests/images/radial_gradients.png")
+        cls.gradients_image.load()
+
+    def test_args_factor(self):
+        im = Image.new("L", (10, 10))
+
+        self.assertEqual((4, 4), im.reduce(3).size)
+        self.assertEqual((4, 10), im.reduce((3, 1)).size)
+        self.assertEqual((10, 4), im.reduce((1, 3)).size)
+
+        with self.assertRaises(ValueError):
+            im.reduce(0)
+        with self.assertRaises(TypeError):
+            im.reduce(2.0)
+        with self.assertRaises(ValueError):
+            im.reduce((0, 10))
+
+    def test_args_box(self):
+        im = Image.new("L", (10, 10))
+
+        self.assertEqual((5, 5), im.reduce(2, (0, 0, 10, 10)).size)
+        self.assertEqual((1, 1), im.reduce(2, (5, 5, 6, 6)).size)
+
+        with self.assertRaises(TypeError):
+            im.reduce(2, "stri")
+        with self.assertRaises(TypeError):
+            im.reduce(2, 2)
+        with self.assertRaises(ValueError):
+            im.reduce(2, (0, 0, 11, 10))
+        with self.assertRaises(ValueError):
+            im.reduce(2, (0, 0, 10, 11))
+        with self.assertRaises(ValueError):
+            im.reduce(2, (-1, 0, 10, 10))
+        with self.assertRaises(ValueError):
+            im.reduce(2, (0, -1, 10, 10))
+        with self.assertRaises(ValueError):
+            im.reduce(2, (0, 5, 10, 5))
+        with self.assertRaises(ValueError):
+            im.reduce(2, (5, 0, 5, 10))
+
+    def test_unsupported_modes(self):
+        im = Image.new("P", (10, 10))
+        with self.assertRaises(ValueError):
+            im.reduce(3)
+
+        im = Image.new("1", (10, 10))
+        with self.assertRaises(ValueError):
+            im.reduce(3)
+
+        im = Image.new("I;16", (10, 10))
+        with self.assertRaises(ValueError):
+            im.reduce(3)
+
+    def get_image(self, mode):
+        mode_info = ImageMode.getmode(mode)
+        if mode_info.basetype == "L":
+            bands = [self.gradients_image]
+            for _ in mode_info.bands[1:]:
+                # rotate previous image
+                band = bands[-1].transpose(Image.ROTATE_90)
+                bands.append(band)
+            # Correct alpha channel to exclude completely transparent pixels.
```

Comment:
"to exclude ~completely~ transparent pixels **completely**"

Reply:
This means "completely transparent pixels should be transformed to not completely transparent by this operation". I'll try to express better.